libraries/freertos_plus/standard/freertos_plus_tcp/source/portable/NetworkInterface/M487/NetworkInterface.c

```diff
@@ -0,0 +1,327 @@
+the size of the stack used by the idle task - but allow this to be overridden in
+FreeRTOSConfig.h as configMINIMAL_STACK_SIZE is a user definable constant. */
+#ifndef configEMAC_TASK_STACK_SIZE
+    #define configEMAC_TASK_STACK_SIZE ( 2 * configMINIMAL_STACK_SIZE )
+#endif
+
+
+static SemaphoreHandle_t xTXMutex = NULL;
+
+/* The handle of the task that processes Rx packets.  The handle is required so
+the task can be notified when new packets arrive. */
+static TaskHandle_t xRxHanderTask = NULL;
+static TimerHandle_t xPhyHandlerTask = NULL;
+/*
+ * A task that processes received frames.
+ */
+static void prvEMACHandlerTask( void *pvParameters );
+static void prvPhyTmrCallback( TimerHandle_t xTimer );
+
+BaseType_t xNetworkInterfaceInitialise( void )
+{
+    uint8_t hwaddr[6];
+    BaseType_t xReturn = pdPASS;
+
+    /* Init ETH */
+    numaker_mac_address(hwaddr);
+    FreeRTOS_UpdateMACAddress(hwaddr);
+    printf("mac address %02x-%02x-%02x-%02x-%02x-%02x \r\n", hwaddr[0], hwaddr[1],hwaddr[2],hwaddr[3],hwaddr[4],hwaddr[5]);
+    /* Enable clock & set EMAC configuration         */
+    /* Enable MAC and DMA transmission and reception */
+    if( numaker_eth_init(hwaddr) < 0)
+    {
+        xReturn = pdFAIL;
+    } else {
+        xReturn = pdPASS;
+        /* Guard against the task being created more than once and the
+        descriptors being initialized more than once. */
+        /* Timer task to monitor PHY Link status */
+        if( xPhyHandlerTask == NULL )
+        {
+            xPhyHandlerTask = xTimerCreate( "TimerPhy",  pdMS_TO_TICKS( 1000 ), pdTRUE, 0, prvPhyTmrCallback );
+            configASSERT(xPhyHandlerTask);
+            xReturn = xTimerStart( xPhyHandlerTask, 0 ) ;
+            configASSERT( xReturn );
+        }
+        /* Rx task */
+        if( xRxHanderTask == NULL )
+        {
+            xReturn = xTaskCreate( prvEMACHandlerTask, "EMAC", configEMAC_TASK_STACK_SIZE, NULL, configMAX_PRIORITIES - 1, &xRxHanderTask );
+            configASSERT( xReturn );
+        }
+        
+        if( xTXMutex == NULL )
+        {
+            xTXMutex = xSemaphoreCreateMutex();
+            configASSERT( xTXMutex );
+        }        
+    }
+
+        NVIC_SetPriority( EMAC_RX_IRQn, configMAC_INTERRUPT_PRIORITY );
+        NVIC_SetPriority( EMAC_TX_IRQn, configMAC_INTERRUPT_PRIORITY );
+
+        numaker_eth_enable_interrupts();
+
+        printf("ETH-RX priority:%d\n",NVIC_GetPriority( EMAC_RX_IRQn));
+
+    return xReturn;
+}
+
+BaseType_t xNetworkInterfaceOutput( NetworkBufferDescriptor_t * const pxDescriptor, BaseType_t xReleaseAfterSend )
+{
+    uint8_t *buffer;
+//    printf("<-- dataLength=%d\n",pxDescriptor->xDataLength);
+    if( pxDescriptor->xDataLength >= PACKET_BUFFER_SIZE )
+    {
+        FreeRTOS_printf(("TX buffer length %d over %d\n", pxDescriptor->xDataLength, PACKET_BUFFER_SIZE));
+        return pdFALSE;
+    }
+    
+    buffer = numaker_eth_get_tx_buf();
```

Comment:
Does this need protection of xTXMutex?

Reply:
@qiutongs  This semaphore is created in xNetworkInterfaceInitialise once. and plus-tcp will not invoke  xNetworkInterfaceInitialise concurrently in multiple thread. So, it's fine.

Reply:
I meant calling function "numaker_eth_get_tx_buf". Is it thread safe?

Reply:
@qiutongs   `numaker_eth_trigger_tx` will dispatch the next tx slot index and it's in semaphore protection. `numaker_eth_get_tx_buf` just check the dispatched slot index and return buffer available or not, so `numaker_eth_get_tx_buf` not need any protection.